src/mixins/createI18nMixin.ts

```diff
@@ -0,0 +1,109 @@
+/* tslint:disable:interface-name */
+import compose, { ComposeFactory } from 'dojo-compose/compose';
+import i18n, { Bundle, getCachedMessages, Messages } from 'dojo-i18n/i18n';
+import { VNodeProperties } from 'maquette';
+import { NodeAttributeFunction, Widget, WidgetOptions, WidgetState } from '../interfaces';
+
+export interface I18nMixin<T extends Messages> {
+	/**
+	 * An array of node attribute functions which return additional attributes that should be mixed into
+	 * the final VNode during a render call. In this particular case, the node's `dir` attribute is optionally
+	 * set when `state.rtl` is a boolean, and the node's `data-locale` attribute is set when `state.locale` is
+	 * not empty.
+	 */
+	nodeAttributes: NodeAttributeFunction[];
+
+	/**
+	 * Return the cached messages for the specified bundle for the current locale, assuming they have already
+	 * benn loaded. If the locale-specific messages have not been loaded, they are fetched and the widget state
+	 * is updated.
+	 *
+	 * @param bundle
+	 * The required bundle object for which available locale messages should be loaded.
+	 */
+	localizeBundle(bundle: Bundle<T>): T;
+}
+
+export interface I18nFactory extends ComposeFactory<I18nMixin<Messages>, WidgetOptions<I18nState>> {}
+
+export interface I18nState extends WidgetState {
+	/**
+	 * The locale for the widget. Is not specified, then the root locale (as determined by `dojo-i18n`) is assumed.
+	 * If specified, the widget's node will have a `data-locale` property set to the locale, in order to facilitate
+	 * styling localized components if the use case arises.
+	 */
+	locale?: string;
+
+	/**
+	 * An optional flag indicating the widget's text direction. If `true`, then the underlying node's `dir`
+	 * property is set to "rtl". If it is `false`, then the `dir` property is set to "ltr". Otherwise, the property
+	 * is not set.
+	 */
+	rtl?: boolean;
+}
+
+/**
+ * @private
+ * An internal helper interface for defining locale and text direction attributes on widget nodes.
+ */
+interface I18nVNodeProperties extends VNodeProperties {
+	'data-locale': string | null;
+	dir: string | null;
+}
+
+export type I18nWidget<M extends Messages, S extends I18nState> = I18nMixin<M> & Widget<S>;
```

Comment:
Since `Widget` only requires a single generic, and since `Messages` is a simple `[key: string]: string` interface, does is it make sense to specify `Messages` as another generic for `I18nWidget`?